pkg/R/sparkRBackend.R

```diff
@@ -0,0 +1,93 @@
+# Methods to call into SparkRBackend. 
+
+
+# Returns TRUE if object is an instance of given class
+isInstanceOf <- function(jobj, className) {
+  stopifnot(class(jobj) == "jobj")
+  cls <- SparkR:::callJStatic("java.lang.Class", "forName", className)
+  SparkR:::callJMethod(cls, "isInstance", jobj)
+}
+
+# Call a Java method named methodName on the object
+# specified by objId. objId should be a "jobj" returned
+# from the SparkRBackend.
+callJMethod <- function(objId, methodName, ...) {
+  stopifnot(class(objId) == "jobj")
+  invokeJava(isStatic=FALSE, objId, methodName, ...)
+}
+
+# Call a static method on a specified className
+callJStatic <- function(className, methodName, ...) {
+  invokeJava(isStatic=TRUE, className, methodName, ...)
+}
+
+# Create a new object of the specified class name
+newJObject <- function(className, ...) {
+  invokeJava(isStatic=TRUE, className, methodName="<init>", ...)
+}
+
+# Remove an object from the SparkR backend. This is done
+# automatically when a jobj is garbage collected.
+removeJObject <- function(objId) {
+  invokeJava(isStatic=TRUE, "SparkRHandler", "rm", objId)
+}
+
+isRemoveMethod <- function(isStatic, objId, methodName) {
+  isStatic == TRUE && objId == "SparkRHandler" && methodName == "rm"
+}
+
+# Invoke a Java method on the SparkR backend. Users
+# should typically use one of the higher level methods like
+# callJMethod, callJStatic etc. instead of using this.
+#
+# If isStatic is true, objId contains className otherwise
+# it should contain a jobj returned previously by the backend
+invokeJava <- function(isStatic, objId, methodName, ...) {
+  if (!exists(".sparkRCon", .sparkREnv)) {
+    stop("No connection to backend found")
+  }
+
+  # If this is already isn't a removeJObject call
+  if (!isRemoveMethod(isStatic, objId, methodName)) {
+    objsToRemove <- ls(.toRemoveJobjs)
+    if (length(objsToRemove) > 0) {
+      #cat("Clearing up objsToRemove\n")
+      sapply(objsToRemove,
+            function(e) {
+              removeJObject(e)
+            })
+      rm(list=objsToRemove, envir=.toRemoveJobjs)
+    }
+  }
+
+  rc <- rawConnection(raw(0), "r+")
+
+  writeBoolean(rc, isStatic)
+  # Write object id as string if it is static
```

Comment:
Not sure I get this comment (e.g. there's no check). Perhaps something like "If static call, objId encodes an object name as string; otherwise write out the object ID"